src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs

```diff
@@ -70,80 +73,58 @@ internal partial class NavigationBarController : ForegroundThreadAffinitizedObje
             presenter.DropDownFocused += OnDropDownFocused;
             presenter.ItemSelected += OnItemSelected;
 
-            subjectBuffer.PostChanged += OnSubjectBufferPostChanged;
-
             // Initialize the tasks to be an empty model so we never have to deal with a null case.
-            _latestModelAndSelectedInfo_OnlyAccessOnUIThread.model = new(
-                ImmutableArray<NavigationBarItem>.Empty,
-                semanticVersionStamp: default,
-                itemService: null!);
+            _latestModelAndSelectedInfo_OnlyAccessOnUIThread.model = new(ImmutableArray<NavigationBarItem>.Empty, itemService: null!);
             _latestModelAndSelectedInfo_OnlyAccessOnUIThread.selectedInfo = new(typeItem: null, memberItem: null);
 
             _modelTask = Task.FromResult(_latestModelAndSelectedInfo_OnlyAccessOnUIThread.model);
+
+            _eventSource = new CompilationAvailableTaggerEventSource(
+                subjectBuffer,
+                asyncListener,
+                TaggerEventSources.OnTextChanged(subjectBuffer),
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer),
+                TaggerEventSources.OnWorkspaceChanged(subjectBuffer, asyncListener),
+                TaggerEventSources.OnWorkspaceRegistrationChanged(subjectBuffer));
```

Comment:
we already have a system for connecting to events and abstracting over them.  This system also handles hooking up buffers to workspaces, making it so that this type doesn't need to track workspaces at all.  Also, a new CompilationAvailableTaggerEventSource source was added as we've seen a bug when roslyn starts where the information can be slightly inaccurate (since we use a frozen-partial snapshot), and we want to move from that to correct data once the compilation is actually available.